### PR TITLE
T9: Add signUp page functionality (#9)

### PR DESCRIPTION
**Pull request by Animansoubi**
add  signUp viewModel and view states
add Firebase dependencies
add  google service json configure
add event activity page and resource
add sign up activity
add auth repository

### CODE REVIEW

File: app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                is SignUpScreenViewState.Error.ShowInvalidPassError -> {
+                    binding.etPass.error = state.errorMessage
+                }
+                is SignUpScreenViewState.Error.ShowEmptyConfirmedPassError -> {
+                    binding.etVerifyPass.error = state.errorMessage
+                }
+                is SignUpScreenViewState.Error.ShowInvalidConfirmedPassError -> {
+                    binding.etVerifyPass.error = state.errorMessage
+                }
+                is SignUpScreenViewState.Error.ShowGeneralError -> {
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt:67]: ❗ Here you need to take an action for showing a general error, maybe showing a toast for now.

File: app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                is SignUpScreenViewState.NavigateToEventScreen -> {
+                    val intent = Intent(this, EventActivity::class.java)
+                    intent.putExtra("extra_object", state.user as Serializable)
+                    startActivity(intent)
+                }
+                is SignUpScreenViewState.SignUpButtonVisibility -> {
+                    binding.btnSignUp.isEnabled = state.isSignUpButtonEnabled
+                }
+                is SignUpScreenViewState.Error.ShowAuthenticationFailedError ->
+                    Toast.makeText(this, state.errorMessage, Toast.LENGTH_LONG).show()
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt:78]: ❗ You have a toast function here then you can create a private showToast() function that takes the text and then show a toast for both places. For general errors and authentication failure errors.

File: app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                is SignUpScreenViewState.Error.ShowEmptyConfirmedPassError -> {
+                    binding.etVerifyPass.error = state.errorMessage
+                }
+                is SignUpScreenViewState.Error.ShowInvalidConfirmedPassError -> {
+                    binding.etVerifyPass.error = state.errorMessage
+                }
+                is SignUpScreenViewState.Error.ShowGeneralError -> {
+                }
+                is SignUpScreenViewState.NavigateToEventScreen -> {
+                    val intent = Intent(this, EventActivity::class.java)
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt:70]: ❔ Do you replace this part with JetPack navigation?

File: app/src/main/java/com/gathering/android/signUp/SignUpScreenViewState.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.gathering.android.signUp.model.User
+
+sealed interface SignUpScreenViewState {
+
+    class NavigateToEventScreen(val user: User) : SignUpScreenViewState
+
+    class SignUpButtonVisibility(val isSignUpButtonEnabled: Boolean) : SignUpScreenViewState
+
+    sealed class Error(val errorMessage: String?) : SignUpScreenViewState {
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/SignUpScreenViewState.kt:11]: Awesome, I liked this part 💪

File: app/src/main/java/com/gathering/android/signUp/SignUpViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    private fun isAllFieldsValid(): Boolean {
+        return isEmailValid &&
+                isPassValid &&
+                isConfirmedPassValid
+    }
+
+    companion object {
+        private const val INVALID_EMail_ADDRESS_FORMAT_ERROR_MESSAGE =
+            "Please enter a valid email address"
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/SignUpViewModel.kt:83]: For now, it's alright but we should remove these strings to XML files later.

File: app/src/main/java/com/gathering/android/signUp/model/AuthRepository.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import com.google.firebase.auth.FirebaseUser
+import com.google.firebase.auth.ktx.auth
+import com.google.firebase.ktx.Firebase
+import javax.inject.Inject
+
+class AuthRepository @Inject constructor() {
+
+    private lateinit var auth: FirebaseAuth
+
+    fun signUpUser(email: String, pass: String, onResponseReady: (ResponseState) -> Unit) {
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/model/AuthRepository.kt:13]: Maybe later we can change this callback to a coroutine, however, it's too much for you. later when you learn that we will migrate to coroutine.

File: app/src/main/java/com/gathering/android/signUp/model/ResponseState.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.gathering.android.signUp.model
+
+sealed interface ResponseState {
+
+    class Success(val user: User) : ResponseState
```
Comment (moyarsuccess) [inline app/src/main/java/com/gathering/android/signUp/model/ResponseState.kt:5]: Good idea to have a proper result from the model layer to view model 💪

**Review by moyarsuccess — APPROVED** [review]

File: app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt (hunk 1 above)
Comment (Animansoubi) [reply app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt:67]: I've just added a toast for general error 👍

File: app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt (hunk 2 above)
Comment (Animansoubi) [reply app/src/main/java/com/gathering/android/signUp/SignUpActivity.kt:78]: I've created a private function for showing the toast :) thanks

File: app/src/main/java/com/gathering/android/signUp/SignUpScreenViewState.kt (hunk 4 above)
Comment (Animansoubi) [reply app/src/main/java/com/gathering/android/signUp/SignUpScreenViewState.kt:11]: thanks for your motivation :)

File: app/src/main/java/com/gathering/android/signUp/SignUpViewModel.kt (hunk 5 above)
Comment (Animansoubi) [reply app/src/main/java/com/gathering/android/signUp/SignUpViewModel.kt:83]: I have some ideas about it but I should double check it with you or @amirziarati

File: app/src/main/java/com/gathering/android/signUp/model/AuthRepository.kt (hunk 6 above)
Comment (Animansoubi) [reply app/src/main/java/com/gathering/android/signUp/model/AuthRepository.kt:13]: I am familiar with coroutine and have applied it to some of my projects but for now, I wanted to make it as simple as possible. for sure we will migrate to coroutine. thank you for  your point

File: app/src/main/java/com/gathering/android/signUp/model/ResponseState.kt (hunk 7 above)
Comment (Animansoubi) [reply app/src/main/java/com/gathering/android/signUp/model/ResponseState.kt:5]: thanks

**Review by moyarsuccess — APPROVED** [review]

**Pull request merged**